### PR TITLE
Add feature to offer to convert code to use deconstructions.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
Feature looks like this:

image

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide review plz

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide review plz

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide review plz

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide review plz

File: src/EditorFeatures/CSharpTest/UseDeconstruction/UseDeconstructionTests.cs [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+        Console.WriteLine(/*7*/t1.name/*8*/ + "" "" + /*9*/t1.age/*10*/);
+    }
+
+    (string name, int age) GetPerson() => default;
+}",
+@"class C
+{
+    void M()
+    {
+        /*1*/(/*2*/int/*3*/ name, /*4*/int/*5*/ age)/*6*/ = GetPerson();
```
Comment (dpoeschl) [inline src/EditorFeatures/CSharpTest/UseDeconstruction/UseDeconstructionTests.cs]: Nice. FYI @chborl

File: src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionCodeFixProvider.cs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+                    variableDeclarator.Initializer.Value),
+                declarationStatement.SemicolonToken);
+        }
+
+        private ExpressionSyntax CreateTupleOrDeclarationExpression(INamedTypeSymbol tupleType, TypeSyntax typeNode)
+        {
+            // If we have an explicit tuple type in code, convert that over to a tuple expression.
+            // i.e.   (int x, int y) t = ...   will be converted to (int x, int y) = ...
+            //
+            // If we had the "var t" form we'll conver that to the declaration expression "var (x, y)"
```
Comment (dpoeschl) [inline src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionCodeFixProvider.cs]: convert

File: src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionCodeFixProvider.cs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                    });
+            }
+
+            return editor.GetChangedRoot();
+        }
+
+        private ForEachVariableStatementSyntax CreateForEachVariableStatement(INamedTypeSymbol tupleType, ForEachStatementSyntax forEachStatement)
+        {
+            // Copy all the tokens/nodes from the existing foreach statement to the new foreach statement.
+            // However, convert the existing declaration over to a "var (x, y)" declaration or (int x, int y)
```
Comment (dpoeschl) [inline src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionCodeFixProvider.cs]: inconsistent quoting

File: src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+            SyntaxNode container, CancellationToken cancellationToken)
+        {
+            var existingSymbols = GetExistingSymbols(semanticModel, container, cancellationToken);
+
+            var reservedNames = semanticModel.LookupSymbols(container.SpanStart)
+                                             .Select(s => s.Name)
+                                             .Concat(existingSymbols.Select(s => s.Name))
+                                             .ToSet();
+
+            foreach (var element in tupleType.TupleElements)
```
Comment (dpoeschl) [inline src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionDiagnosticAnalyzer.cs]: expanded for perf?

File: src/Workspaces/Core/Portable/Shared/Extensions/SyntaxEditorExtensions.cs [hunk 5]
```diff
@@ -14,6 +14,48 @@ namespace Microsoft.CodeAnalysis.Shared.Extensions
     internal static class SyntaxEditorExtensions
     {
         /// <summary>
+        /// Performs several edits do a document.  If multiple edits are made within the same
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/Shared/Extensions/SyntaxEditorExtensions.cs]: to a document

File: src/Workspaces/Core/Portable/Shared/Extensions/SyntaxEditorExtensions.cs [hunk 6]
```diff
@@ -14,6 +14,48 @@ namespace Microsoft.CodeAnalysis.Shared.Extensions
...
+                editor, document,
+                originalNodes,
+                (syntaxFacts, node) => GetExpressionSemanticBoundary(syntaxFacts, node),
+                canReplace,
+                updateRoot,
+                cancellationToken);
+        }
+
+        /// <summary>
+        /// Performs several edits do a document.  If multiple edits are made within a method
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/Shared/Extensions/SyntaxEditorExtensions.cs]: to a document

File: src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionDiagnosticAnalyzer.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionDiagnosticAnalyzer.cs]: yup.  but i don't feel strongly about it.

File: src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionDiagnosticAnalyzer.cs [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+            var localName = local.Name;
+
+            foreach (var identifierName in searchScope.DescendantNodes().OfType<IdentifierNameSyntax>())
+            {
+                if (identifierName.Identifier.ValueText == localName)
+                {
+                    var symbol = semanticModel.GetSymbolInfo(identifierName, cancellationToken).GetAnySymbol();
+                    if (local.Equals(symbol))
+                    {
+                        if (!(identifierName.Parent is MemberAccessExpressionSyntax memberAccess))
```
Comment (rchande) [inline src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionDiagnosticAnalyzer.cs]: What if we're inside a for loop that declares a local whose type name is the same as the tuple variable's name? Isn't the `Program` in the loop an IdentiferNameSYntax?
```
    class Program
    {
        static void Main(string[] args)
        {
            foreach ((int, int) Program in GetZ())
            {
                Program p = null;
                Console.WriteLine(p);
            }
        }

        static IEnumerable<(int, int)> GetZ() { }
    }
```

Comment (rchande) [reply src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionDiagnosticAnalyzer.cs]: (yes, very contrived)

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionDiagnosticAnalyzer.cs]: I'm not sure i understand your example.  The purpose of this feature is to see how the tuple-typed variable is used.  i.e. as long as you only use it to access tuple fields, then it's safe to elide the tuple altogether.  But, if you access things like methods on the tuple, or you use the variable in any other manner, then we assume it's not safe to use.

In your example I'm not sure what's going on (as you have a type and variable with teh same name).

Comment (rchande) [reply src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionDiagnosticAnalyzer.cs]: Oh, never mind--I can read.

**Comment by rchande** [comment]
LGTM but we (I?) should probably test this out before we merge it.

**Comment by CyrusNajmabadi** [comment]
Oh yeah, i def think this should be tested out. 

Note: one open design concern is that this probably is *not* a good idea for something like a large tuple.  if someone is returning 8 values, do we really want to explode that in the code?  

We may want to consider making this On|None.  That way you can easily invoke this, but you don't get prompted to make the change if it's not something you don't want to do.

**Comment by rchande** [comment]
Played with this some and it seems safe to merge--I'll add it to our test plan to make sure it gets depth testing as well.

**Comment by dpoeschl** [comment]
retest perf_correctness_prtest please

**Comment by dpoeschl** [comment]
@TyOverby Another null ref in perf_correctness in https://ci.dot.net/job/dotnet_roslyn/job/master/job/perf_correctness_prtest/9445/

**Comment by dpoeschl** [comment]
retest windows_debug_unit32_prtest please

**Comment by dpoeschl** [comment]
@Pilchie What do you think about this one? It has two approvals and Ravi tested it out locally. I think the only open question is whether it's desirable on large tuples, but the feature is disabled by default and we can get feedback from those who want to try it out.

Submitting for Ask Mode.

**Comment by CyrusNajmabadi** [comment]
> , but the feature is disabled by default

@dpoeschl Actually, it's currently "TrueWithSuggestionEnforcement".  

I'm leaning toward that being the right choice.  I think it should be pretty rare for people to really massively overload tuples.  Becaudse then you end up with something awful like:

```c#
private (string firstName, string middleName, string lastName, int age, string address, string address2, string zipCode, string countryCode) GetPerson()
```

And everything is already awful.  

I think defaulting to suggestions here is ok.  For the vast majority of cases expanding will likely be a good thing.  In the rare case you don't want it, just don't invoke it.  And for people who really don't like the expanded form and prefer "localTuple.someField", they can just disable.  

That said, i don't feel strongly.  If there is any concern, setting it to "TrueWithNone" would be fine with me.

Thoughts?

**Comment by dpoeschl** [comment]
Talked with @kuhlenh and we think it should be fine as a suggestion as it is a new language feature that people are still figuring out, and this will almost always help them do the simpler thing.

**Comment by CyrusNajmabadi** [comment]
Cool!  Thanks!

**Comment by Pilchie** [comment]
Treating @rchande's comments as a code review approval, I'm good with this.

**Comment by CyrusNajmabadi** [comment]
Thanks guys!

**Pull request merged**